registry/remote/repository.go

```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+		// expensive calculation
+		if calculatedDigest, err = calculateDigestFromResponse(resp); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		contentDigest = calculatedDigest
+	} else {
+		contentDigest = hdrDigest
 	}
 
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	// 5b. With all 3 digests appropriately set, perform final 3-way validation
+	if err = verifyContentDigests(calculatedDigest, hdrDigest, refDigest); err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("digest validation failure; %w", err)
 	}
 
+	// 6. Finally, if we made it this far, then all is good; return.
 	return ocispec.Descriptor{
 		MediaType: mediaType,
 		Digest:    contentDigest,
 		Size:      size,
 	}, nil
 }
 
+// calculateDigestFromResponse calculates the actual digest of the response body
+// taking care not to destroy it in the process.
+func calculateDigestFromResponse(resp *http.Response) (digest.Digest, error) {
+	// Read from `resp.Body` and (via io.TeeReader), duplicate the read data
+	// into the new buffers.  This read will "deplete" resp.Body which, having
+	// no "rewind" button, will become useless.  At which point, one clone is
+	//used to calculate the digest, and another is used replace the now-defunct
+	// `resp.Body`.
+	var cloneForDigestCalculation io.Reader
+	var cloneForRestoration bytes.Buffer
+
+	var d digest.Digest
+	var err error
+	body := limitReader(resp.Body, 0)
+	cloneForDigestCalculation = io.TeeReader(body, &cloneForRestoration)
+	if d, err = digest.FromReader(cloneForDigestCalculation); err != nil {
+		return d, fmt.Errorf("%s %q: failed to read response body: %w", resp.Request.Method, resp.Request.URL, err)
+	}
+
+	resp.Body = io.NopCloser(&cloneForRestoration)
+
+	return d, nil
+}
+
 // verifyContentDigest verifies "Docker-Content-Digest" header if present.
 // OCI distribution-spec states the Docker-Content-Digest header is optional.
 // Reference: https://github.com/opencontainers/distribution-spec/blob/v1.0.1/spec.md#legacy-docker-support-http-headers
 func verifyContentDigest(resp *http.Response, expected digest.Digest) error {
-	digestStr := resp.Header.Get("Docker-Content-Digest")
+	digestStr := resp.Header.Get(dockerContentDigestHeader)
 	if digestStr == "" {
 		return nil
 	}
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
-	}
-	if contentDigest != expected {
+	if contentDigest, err := digest.Parse(digestStr); err != nil {
+		return fmt.Errorf(
+			"%s %q: invalid response %s: %s",
+			resp.Request.Method, resp.Request.URL, dockerContentDigestHeader, digestStr,
+		)
+	} else if contentDigest != expected {
```

Comment:
```suggestion
	}
	if contentDigest != expected {
```
See https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow